modules/grandpa/src/lib.rs

```diff
@@ -107,7 +133,10 @@ pub mod pallet {
 		///
 		/// If successful in verification, it will write the target header to the underlying storage
 		/// pallet.
-		#[pallet::weight(0)]
+		#[pallet::weight(T::WeightInfo::submit_finality_proof(
+			T::BridgedSessionLength::get().as_() as u32,
+			T::BridgedValidatorCount::get() as u32,
```

Comment:
I think we can be much more accurate with the validator count and get rid of this hardcoded parameter.
Either we can read it from the storage (I don't know the exact rules about weights, but it's probably discouraged due to extra state read) or we can have the submitter provide the current validator set length for more accurate weight calculation and we simply compare that with an actual validator set.

Happy to work on this under separate issue though as a weight cost optimisation, since it requires a change in the relayer code as well.

Reply:
> Either we can read it from the storage (I don't know the exact rules about weights, but it's probably discouraged due to extra state read) 

Yeah, we shouldn't be reading from storage in order to calculate weights.

> Happy to work on this under separate issue though as a weight cost optimisation, since it requires a change in the relayer code as well.

👍